staging/src/k8s.io/client-go/tools/portforward/portforward.go

```diff
@@ -184,10 +246,18 @@ func (pf *PortForwarder) forward() error {
 // listenOnPort delegates tcp4 and tcp6 listener creation and waits for connections on both of these addresses.
 // If both listener creation fail, an error is raised.
 func (pf *PortForwarder) listenOnPort(port *ForwardedPort) error {
-	errTcp4 := pf.listenOnPortAndAddress(port, "tcp4", "127.0.0.1")
-	errTcp6 := pf.listenOnPortAndAddress(port, "tcp6", "::1")
-	if errTcp4 != nil && errTcp6 != nil {
-		return fmt.Errorf("All listeners failed to create with the following errors: %s, %s", errTcp4, errTcp6)
+	var listenSuccessfull bool
+	var errors []error
+	for _, addr := range pf.addresses {
```

Comment:
silently failing on some of the inputs is not what I would expect. I'm unsure whether we want to support the user specifying multiple addresses to bind to... would we expect the command to exit if any were unsuccessful, or if all were unsuccessful? I could see both ways, which makes me think it might be better to limit a single invocation to a single address.

Reply:
at it's current implementation, it binds to two addresses (127.0.0.1, ::1) by default so that behavior should be kept, it is in my opinion perfectly logical to want to listen on particular set of IP addresses and not the others (ie. IP from one external interface and the loopback ones). The command in kubectl explicitly lists ports it successfully bound to

Reply:
> at it's current implementation, it binds to two addresses (127.0.0.1, ::1) by default so that behavior should be kept

sure, if the user doesn't have an address they want to forward on, it should continue binding to "local" addresses.

> it is in my opinion perfectly logical to want to listen on particular set of IP addresses and not the others (ie. IP from one external interface and the loopback ones)

I agree, but if I explicitly want to listen on a set of IP addresses, I would not expect the command to report success if it couldn't accomplish that. Limiting the user-specified bind address to a single address makes it super clear when their intent was not satisfied, and plays well with something using this to script port-forward setups (e.g. loops that restart port-forward on a particular address when it times out or dies)

Reply:
> The argument for scripting "compatibility" makes a lot of sense. I would lean to the idea that it could be implemented with failing if any bind failed

There's the initial bind failure to consider, but also what happens when one of the connections closes, right? I'd expect the following:
* On start, all explicitly requested bind addresses must succeed
* On close of any explicitly requested bind addresses, the entire command should exit

I would stick with the user specifying addresses explicitly (not doing localhost->[127.0.0.1, ::1] expansion for explicit addresses). You can move the defaulting into the command if it makes it easier to distinguish between unspecified addresses (which should preserve best-effort binding to 127.0.0.1 and ::1) and explicitly specified addresses of `127.0.0.1,::1` (which should both succeed or it would be considered failure)

Reply:
I was looking into it a bit more, and it seems in fact that port-forward is not really built to behave like this at the moment. Look at following vanilla kubectl port-forward : 

    $ kubectl port-forward mysql-server-0 3306:3306 80:3306  
    Forwarding from 127.0.0.1:3306 -> 3306
    Unable to listen on port 80: All listeners failed to create with the following errors: Unable to create listener: Error listen tcp4 127.0.0.1:80: bind: permission denied, Unable to create listener: Error listen tcp6: address [[::1]]:80: missing port in address

So... as long as one of the ports binds correctly, the port-forward stays up and running. 

I'm not saying that is the ultimate argument for not failing if only some of the addresses fail to bind, but to stay consistent, it would stand to reason that we change to fail if any port failed as well, which would be a change from current behavior.

Reply:
What do you think about changing it to just offer a single address? Users can always run multiple instances of kubectl, and that way they if one fails the user can restart it.

Reply:
I see there's a good case for running on all loopback addresses, at least.

In that case, I think it's super important that if it can't bind every address requested, it fail-- if something malicious is occupying one of the ports, it'd be a security vuln.

Reply:
as mentioned before, one of the cases where we will fail on one, but not all, is when we make a `localhost` bind on a system that does not support IPv6

Reply:
Since this PR isn't making things worse, maybe we should merge this and file an issue for followup.